File: src/Tests/Behavioral/SubpackageFuncTypeParam/SubpackageFuncTypeParam.go
// Guards a methodless named func type whose parameter references a SUBPACKAGE type — an import path
// with a slash, e.g. sync/atomic. When the func type collapses to its base C# delegate, the parameter
// type is rendered from the signature's t.String() (`func(*sync/atomic.Int32) int32`), which lost the
// file import alias. The subpackage path `sync/atomic` must resolve to the package CLASS
// `sync.atomic_package`, NOT the namespace-dotted `sync.atomic` (CS0234 — `atomic` is not a namespace
// of `go.sync`). Same shape as path/filepath's WalkDirFunc referencing io/fs.DirEntry.
package main

import (
	"fmt"
	"sync/atomic"
)

// methodless named func type with a subpackage-type (*atomic.Int32) parameter.
type applyFunc func(c *atomic.Int32) int32

func run(f applyFunc, c *atomic.Int32) int32 {
	return f(c)
}

func main() {
	var c atomic.Int32
	c.Store(5)
	fmt.Println(run(func(c *atomic.Int32) int32 { return c.Load() }, &c)) // 5
	c.Add(10)
	fmt.Println(run(func(c *atomic.Int32) int32 { return c.Load() }, &c)) // 15
}

File: src/Tests/Behavioral/SubpackageFuncTypeParam/SubpackageFuncTypeParam.cs
namespace go;

using fmt = fmt_package;
using atomic = sync.atomic_package;
using sync;

partial class main_package {

// type applyFunc is a methodless func type — rendered inline as its base delegate

internal static int32 run(Func<ж<atomic.Int32>, int32> f, ж<atomic.Int32> Ꮡc) {
    ref var c = ref Ꮡc.Value;

    return f(Ꮡc);
}

internal static void Main() {
    ref var c = ref heap(new atomic.Int32(), out var Ꮡc);
    Ꮡc.Store(5);
    fmt.Println(run((ж<atomic.Int32> cΔ1) => cΔ1.Load(), Ꮡc));
    Ꮡc.Add(10);
    fmt.Println(run((ж<atomic.Int32> cΔ2) => cΔ2.Load(), Ꮡc));
}

} // end main_package
